STR-1297: Make ASM state progress within CSM machinery

evgenyzdanovich opened the pull request:
## Description

<!--
Provide a brief summary of the changes and the motivation behind them.
-->

initial **ASM** integration. 

**Includes**:

- implement ASM as a service.
- run ASM service in `strata-client` alongside existing CSM.
- define RocksDB and SLED tables for ASM persistence.
- define minimal interface to notify subprotocols (via subprotocol service handle).

**Doesn't include**:

- versioning
- storage of aux inputs

### Type of Change

<!--
Select the type of change your PR introduces (put an `x` in all that apply):
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactor
- [ ] New or updated tests
- [ ] Dependency Update

## Notes to Reviewers

<!--
Anything in particular you want to note that will help reviewers fulfill their role
in reviewing this PR?
-->

## Checklist

<!--
Ensure all the following are checked:
-->

- [x] I have performed a self-review of my code.
- [x] I have commented my code where necessary.
- [x] I have updated the documentation if needed.
- [x] My changes do not introduce new warnings.
- [x] I have added (where necessary) tests that prove my changes are effective or that my feature works.
- [x] New and existing tests pass with my changes.
- [ ] I have [disclosed my use of AI](https://github.com/alpenlabs/alpen/blob/main/.github/CONTRIBUTING.md#ai-assistance-notice)
      in the body of this PR.

## Related Issues

<!--
Link any related issues (e.g., `closes #123`, `fixes #456`).
-->

prajwolrg commented on crates/asm/common/src/state.rs:
```diff
@@ -12,7 +12,7 @@ use crate::{AsmError, Mismatched, Subprotocol, SubprotocolId};
 /// receiving protocol transactions at L1 and updating its storage. A zk-SNARK proof
 /// attests that the transition from the previous ASM state to the new state
 /// was performed correctly on the given L1 block.
-#[derive(Clone, Debug, BorshSerialize, BorshDeserialize)]
+#[derive(Clone, Debug, BorshSerialize, BorshDeserialize, PartialEq)]
```
nit (derive order): use built-in derives first before those that come from external crates.

delbonis reviewed (changes requested):
This is really unfortunate, there's been some serious miscommunication here, I'm not sure where that happened.

At this level, subprotocols are completely encapsulated within the ASM infrastructure.  CSM doesn't need to be aware of them, they are not separate workers, so the `CommandHandle` doesn't exist, there is not a separate ASM worker.  There doesn't need to be a separate `AsmManager` or ASM DB types because it should be a field inside the CSM state.

The way it should have worked it that the CSM state struct would be extended like this:

```rs
pub struct ClientState {
    // Last *finalized* checkpoint.
    pub(crate) last_finalized_checkpoint: Option<L1Checkpoint>,

    // Last *seen* checkpoint.
    pub(crate) last_seen_checkpoint: Option<L1Checkpoint>,

    /// ASM state computed as of this block.
    pub(crate) asm_state: AnchorState, // sorry the name is weird, we could change that
}
```

And then most of the job of the CSM STF is:

* looking at the old state, figuring out what aux input there needs to be ("preprocessing", probably empty for now)
* invoking the ASM with that (`compute_asm_transition`)
* responding to emitted logs to update checkpoint fields and maybe other things in the future

Then the CSM STF output which gets persisted would include the new CSM state (containing the new ASM state) and the ASM aux input (which we'll later use for generating ASM proofs).

So this PR has a lot more moving parts than was necessary.

evgenyzdanovich commented:
@delbonis I was under impression that it's better to create a separate ASM worker because it seems that part of CSM (specifically, `ClientState` and its stf) will eventually be replaced by Checkpoint Subprotocol (and thus, we will need some other suitable "driver"). 
Also, the notion of "following l1 chain" is kinda duplicated as well (in both the CSM and ASM worker).

Once we are ready, we just create a (thin) subprotocol service that will do "responding to emitted logs to update checkpoint fields/ DB/signal on the status channel/communication to FCM/etc". This subprotocol service will be registered and data to it is passed by this ASM worker service.

P.S. Current impl already supports registering subprotocol handles(`crates/asm/worker/src/builder.rs`) where logs/state are passed to subprotocol service/task.


-------------------
A random philosophical note. 
The approach I had in mind allows us building from the ground up and switch parts along the way (but there can be some boilerplate/duplication). 
If I understand you correctly now, what you are proposing is to start growing ASM more tighly coupled to CSM and do more like refactors along the way (which, IMO, is a trickier thing to do).

Not a strong proponent of any approach, as long as we agree on it within the team (and ideally before impl - although that's not always possible:)

bewakes reviewed (changes requested):
I was also under the assumption that there would be csm re-structured as a service and that it would internally run ASM stf.

My main suggestions are regarding units tests that some of the important modules are missing.

evgenyzdanovich commented on crates/state/src/asm_state.rs:
```diff
@@ -0,0 +1,31 @@
+use strata_asm_common::{AnchorState, AsmLogEntry};
+use strata_asm_stf::AsmStfOutput;
+#[derive(Debug, Clone, BorshSerialize, BorshDeserialize, PartialEq)]
+pub struct AsmState {
+    state: AnchorState,
+    logs: Vec<AsmLogEntry>,
+}
```
Well, that's just an "umbrella" state handy for bookkeeping / passing data around

evgenyzdanovich commented on crates/db-store-rocksdb/src/asm/db.rs:
```diff
@@ -0,0 +1,54 @@
+
+#[cfg(feature = "test_utils")]
+#[cfg(test)]
+mod tests {}
```
done

evgenyzdanovich commented on crates/db-store-sled/src/asm/db.rs:
```diff
@@ -0,0 +1,29 @@
+        let res = iter.next();
+        Ok(res.transpose()?)
+    }
+}
```
done

evgenyzdanovich commented on crates/db-store-sled/src/asm/schemas.rs:
```diff
@@ -0,0 +1,10 @@
+use strata_primitives::l1::L1BlockCommitment;
+use strata_state::asm_state::AsmState;
+
+use crate::{define_table_with_seek_key_codec, define_table_without_codec, impl_borsh_value_codec};
+
+// ASM state per block schema and corresponding codecs implementation
+define_table_with_seek_key_codec!(
+    /// A table to store ASM state per l1 block.
+    (AsmBlockSchema) L1BlockCommitment => AsmState
+);
```
separating schemas is fine, done.

evgenyzdanovich commented:
@delbonis I re-read your "New service primitive" doc and modified the ASM service's status (also removed explicit subprotocol handlers), so that async code can listen to ASM log/state updates (as specified in the doc).

delbonis commented on crates/asm/worker/src/handle.rs:
```diff
@@ -1,24 +1,39 @@
-pub struct AsmWorkerHandle {
+pub struct AsmWorkerHandle<W: WorkerContext + Send + Sync + 'static> {
     command_handle: CommandHandle<L1BlockCommitment>,
+    service_monitor: Monitor<W>,
```
This is super ugly, I should have converted this like I did the command handle type to just use the status type, so that the monitor type doesn't have to itself be generic over the context (and make all the consumers be aware of the concrete context type).  It's fun and quirky to have it be generic over the service but it's actually cumbersome when the service itself is generic.  Maybe I should have reworked how we'd use a context trait instead, but that sounds like it's getting close to needing HKTs.

evgenyzdanovich replied:
I vaguely plan on working around "integrating CSM with ASM logs" later, so I'll do how it's done in CommandHandle, this will improve the ergonomics a bit.

delbonis commented on crates/state/src/asm_state.rs:
```diff
@@ -31,3 +33,33 @@ impl AsmState {
     }
 }
+
+/// Serde serialization of logs as hex strings.
+mod asm_log_entry_vec_hex {
```
Is it not enough to add some attrs to `AsmLogEntry`?

evgenyzdanovich replied:
yeah, added.

delbonis commented:
@evgenyzdanovich 

I'm not against rewriting "the stuff that will go around the ASM STF invocation itself" entirely.  I can see the value in having a more clean foundation to build on top of.

> Logs are stored in DB and interpreted by the subprotocol service logic (additional, separate async task alongside subprotocol state which is internal to ASM). 

I agree we shouldn't be interpreting things like bridge subproto logs in this service.  But at the very least we'd need to consume the vk change upgrade log and record that outside of the `AnchorState` type itself to correctly control execution going forwards.  So that's why I was pointing towards reusing/extending this here.  I believe there would be similar bookkeeping needs for how we'd eventually, although I am not sure where that left off after the initial implementation.

So my thinking was if we're going to have that anyways, we might as well use it for other essential components like taking the checkpoint subproto logs and remember those as part of the same lifecycle, so we can make it available and not need to run another service just to track that.

The bridge related messages it would make more sense to run as a separate service because it's more "on the side", so I concur on that point.

> But logic paths are different oftentimes: for example btcio internals are different (CSM currently consumes L1BlockManifest which is in an old format, ASM consumes pure BTC blocks and does parsing of sps-50 under the hood).

Yeah this does have some messiness around the filters that are irrelevant now.

However, I had suggested that we should maintain this L1BlockManifest type but just incrementally convert it over to having SPS-50 tagged txs.  I had originally drafted all the spec docs for the ASM around taking full blocks but I think it makes more sense in practice to separate it out into two phases, one where we scrape the txs (is really fast and doesn't need the reset of the CSM state) and then the second where we really invoke the main ASM execution.  This would necessitate the proof being two steps.

> Sorry, I don't see any coupling here, could you please elaborate? ASM-as-a-service runs its STF (consuming pure btc blocks) and storing/making an output opaque asm logs.

Maybe this was changed but I was seeing what looked like "subprotocol command handles".  I interpreted this as having a service for each subprotocol that responds to the logs it produces.

> in this discussion I frankly hold only one strong opinion - ideally I'd like to see CSM being a service eventually in the end.

Completely agree here.

delbonis reviewed (changes requested):
^

evgenyzdanovich commented:
@delbonis thanks for the feedback.

>But at the very least we'd need to consume the vk change upgrade log and record that outside of the AnchorState type itself to correctly control execution going forwards.
Agree. This is pretty much the first impl, we'll iterate on that soon.

>So my thinking was if we're going to have that anyways, we might as well use it for other essential components like taking the checkpoint subproto logs and remember those as part of the same lifecycle, so we can make it available and not need to run another service just to track that.
well, this is (now) makes sense to me. I think in the (near) future we can end up having what you described (in some shape).

> Maybe this was changed but I was seeing what looked like "subprotocol command handles". I interpreted this as having a service for each subprotocol that responds to the logs it produces.

Yeah, I removed it (well, after re-reading your initial doc) and made sure the data is available through service monitor.

evgenyzdanovich commented on crates/asm/worker/src/state.rs:
```diff
@@ -0,0 +1,147 @@
+    /// ASM spec for ASM STF.
+    asm_spec: StrataAsmSpec,
+
+    /// A runtime handle.
+    _runtime_handle: Handle,
```
done

evgenyzdanovich commented on crates/storage/src/lib.rs:
```diff
@@ -20,6 +20,7 @@ use strata_db::traits::DatabaseBackend;
 #[derive(Clone)]
 #[expect(missing_debug_implementations)]
 pub struct NodeStorage {
+    asm_manager: Arc<AsmManager>,
```
done

evgenyzdanovich commented:
> needing wrapping container to track STF version, remembering computed aux inputs, etc.

Thanks, I left a TODO to address that. It's much easier to work on separately later. Right now, the main objective is to fully integrate checkpoint v0 with csm (which doesn't require neither of those things) and the current impl roughly suffice.

I'd much rather prefer to shift much of our discussions to next PR(s) and have polishing there, but merge this one soon(er) so we can start working on the actual integration (CSM consuming ASM logs).
